// @insforge/shared-schemas/dist/error-codes.js
// Canonical error codes shared across backend, SDK, CLI, MCP, and tooling.
// Keep the public string values stable.
export const ERROR_CODES = {
    // AUTH module
    AUTH_INVALID_EMAIL: 'AUTH_INVALID_EMAIL',
    AUTH_WEAK_PASSWORD: 'AUTH_WEAK_PASSWORD',
    AUTH_INVALID_CREDENTIALS: 'AUTH_INVALID_CREDENTIALS',
    AUTH_INVALID_API_KEY: 'AUTH_INVALID_API_KEY',
    AUTH_EMAIL_EXISTS: 'AUTH_EMAIL_EXISTS',
    AUTH_OAUTH_CONFIG_ERROR: 'AUTH_OAUTH_CONFIG_ERROR',
    AUTH_UNSUPPORTED_PROVIDER: 'AUTH_UNSUPPORTED_PROVIDER',
    AUTH_TOKEN_EXPIRED: 'AUTH_TOKEN_EXPIRED',
    AUTH_UNAUTHORIZED: 'AUTH_UNAUTHORIZED',
    AUTH_NEED_VERIFICATION: 'AUTH_NEED_VERIFICATION',
    AUTH_SIGNUP_DISABLED: 'AUTH_SIGNUP_DISABLED',
    // DATABASE module
    DATABASE_INVALID_PARAMETER: 'DATABASE_INVALID_PARAMETER',
    DATABASE_VALIDATION_ERROR: 'DATABASE_VALIDATION_ERROR',
    DATABASE_CONSTRAINT_VIOLATION: 'DATABASE_CONSTRAINT_VIOLATION',
    DATABASE_NOT_FOUND: 'DATABASE_NOT_FOUND',
    DATABASE_DUPLICATE: 'DATABASE_DUPLICATE',
    DATABASE_PERMISSION_DENIED: 'DATABASE_PERMISSION_DENIED',
    DATABASE_INTERNAL_ERROR: 'DATABASE_INTERNAL_ERROR',
    DATABASE_FORBIDDEN: 'DATABASE_FORBIDDEN',
    // STORAGE module
    STORAGE_INVALID_PARAMETER: 'STORAGE_INVALID_PARAMETER',
    STORAGE_INVALID_FILE_TYPE: 'STORAGE_INVALID_FILE_TYPE',
    STORAGE_INSUFFICIENT_QUOTA: 'STORAGE_INSUFFICIENT_QUOTA',
    STORAGE_NOT_FOUND: 'STORAGE_NOT_FOUND',
    STORAGE_PERMISSION_DENIED: 'STORAGE_PERMISSION_DENIED',
    // STORAGE module — S3 gateway
    S3_ACCESS_KEY_LIMIT_EXCEEDED: 'S3_ACCESS_KEY_LIMIT_EXCEEDED',
    S3_ACCESS_KEY_NOT_FOUND: 'S3_ACCESS_KEY_NOT_FOUND',
    S3_PROTOCOL_UNAVAILABLE: 'S3_PROTOCOL_UNAVAILABLE',
    // REALTIME module
    REALTIME_CONNECTION_FAILED: 'REALTIME_CONNECTION_FAILED',
    REALTIME_UNAUTHORIZED: 'REALTIME_UNAUTHORIZED',
    REALTIME_INVALID_EVENT: 'REALTIME_INVALID_EVENT',
    // AI module
    AI_INVALID_API_KEY: 'AI_INVALID_API_KEY',
    AI_INVALID_MODEL: 'AI_INVALID_MODEL',
    AI_UPSTREAM_UNAVAILABLE: 'AI_UPSTREAM_UNAVAILABLE',
    // LOGS module
    LOGS_AWS_NOT_CONFIGURED: 'LOGS_AWS_NOT_CONFIGURED',
    // COMPUTE module
    COMPUTE_CLOUD_UNAVAILABLE: 'COMPUTE_CLOUD_UNAVAILABLE',
    COMPUTE_NOT_CONFIGURED: 'COMPUTE_NOT_CONFIGURED',
    COMPUTE_PROVIDER_ERROR: 'COMPUTE_PROVIDER_ERROR',
    COMPUTE_SERVICE_NOT_FOUND: 'COMPUTE_SERVICE_NOT_FOUND',
    COMPUTE_SERVICE_NOT_CONFIGURED: 'COMPUTE_SERVICE_NOT_CONFIGURED',
    COMPUTE_SERVICE_DEPLOY_FAILED: 'COMPUTE_SERVICE_DEPLOY_FAILED',
    COMPUTE_SERVICE_ALREADY_EXISTS: 'COMPUTE_SERVICE_ALREADY_EXISTS',
    COMPUTE_SERVICE_START_FAILED: 'COMPUTE_SERVICE_START_FAILED',
    COMPUTE_SERVICE_STOP_FAILED: 'COMPUTE_SERVICE_STOP_FAILED',
    COMPUTE_SERVICE_DELETE_FAILED: 'COMPUTE_SERVICE_DELETE_FAILED',
    COMPUTE_REGION_CHANGE_NOT_SUPPORTED: 'COMPUTE_REGION_CHANGE_NOT_SUPPORTED',
    COMPUTE_QUOTA_EXCEEDED: 'COMPUTE_QUOTA_EXCEEDED',
    // Billing module
    BILLING_INSUFFICIENT_BALANCE: 'BILLING_INSUFFICIENT_BALANCE',
    // EMAIL module
    EMAIL_SMTP_CONNECTION_FAILED: 'EMAIL_SMTP_CONNECTION_FAILED',
    EMAIL_SMTP_SEND_FAILED: 'EMAIL_SMTP_SEND_FAILED',
    EMAIL_TEMPLATE_NOT_FOUND: 'EMAIL_TEMPLATE_NOT_FOUND',
    // Deployments module
    DEPLOYMENT_ALREADY_EXISTS: 'DEPLOYMENT_ALREADY_EXISTS',
    DEPLOYMENT_INVALID_FILE: 'DEPLOYMENT_INVALID_FILE',
    DEPLOYMENT_NOT_FOUND: 'DEPLOYMENT_NOT_FOUND',
    DOMAIN_ALREADY_EXISTS: 'DOMAIN_ALREADY_EXISTS',
    DOMAIN_INVALID: 'DOMAIN_INVALID',
    DOMAIN_NOT_FOUND: 'DOMAIN_NOT_FOUND',
    ENVIRONMENT_VARIABLE_NOT_FOUND: 'ENVIRONMENT_VARIABLE_NOT_FOUND',
    // Schedules module
    SCHEDULE_INVALID_CRON: 'SCHEDULE_INVALID_CRON',
    SCHEDULE_NOT_FOUND: 'SCHEDULE_NOT_FOUND',
    // Payments module
    PAYMENT_CONFIG_INVALID: 'PAYMENT_CONFIG_INVALID',
    PAYMENT_NOT_FOUND: 'PAYMENT_NOT_FOUND',
    PAYMENT_METHOD_DECLINED: 'PAYMENT_METHOD_DECLINED',
    // Secrets module
    SECRET_ALREADY_EXISTS: 'SECRET_ALREADY_EXISTS',
    SECRET_NOT_FOUND: 'SECRET_NOT_FOUND',
    // General
    MISSING_FIELD: 'MISSING_FIELD',
    ALREADY_EXISTS: 'ALREADY_EXISTS',
    INVALID_INPUT: 'INVALID_INPUT',
    NOT_FOUND: 'NOT_FOUND',
    UNKNOWN_ERROR: 'UNKNOWN_ERROR',
    INTERNAL_ERROR: 'INTERNAL_ERROR',
    TOO_MANY_REQUESTS: 'TOO_MANY_REQUESTS',
    FORBIDDEN: 'FORBIDDEN',
    RATE_LIMITED: 'RATE_LIMITED',
};
