// @insforge/shared-schemas/dist/payments.schema.js
import { z } from 'zod';
export const stripeEnvironmentSchema = z.enum(['test', 'live']);
export const stripeConnectionStatusSchema = z.enum(['unconfigured', 'connected', 'error']);
export const stripeLatestSyncStatusSchema = z.enum(['succeeded', 'failed']);
export const stripeConnectionSchema = z.object({
    environment: stripeEnvironmentSchema,
    status: stripeConnectionStatusSchema,
    stripeAccountId: z.string().nullable(),
    stripeAccountEmail: z.string().nullable(),
    accountLivemode: z.boolean().nullable(),
    webhookEndpointId: z.string().nullable(),
    webhookEndpointUrl: z.string().nullable(),
    webhookConfiguredAt: z.string().nullable(),
    maskedKey: z.string().nullable(),
    lastSyncedAt: z.string().nullable(),
    lastSyncStatus: stripeLatestSyncStatusSchema.nullable(),
    lastSyncError: z.string().nullable(),
    lastSyncCounts: z.record(z.number()),
});
export const stripeProductSchema = z.object({
    environment: stripeEnvironmentSchema,
    stripeProductId: z.string(),
    name: z.string(),
    description: z.string().nullable(),
    active: z.boolean(),
    defaultPriceId: z.string().nullable(),
    metadata: z.record(z.string()),
    syncedAt: z.string(),
});
export const stripePriceSchema = z.object({
    environment: stripeEnvironmentSchema,
    stripePriceId: z.string(),
    stripeProductId: z.string().nullable(),
    active: z.boolean(),
    currency: z.string(),
    unitAmount: z.number().nullable(),
    unitAmountDecimal: z.string().nullable(),
    type: z.string(),
    lookupKey: z.string().nullable(),
    billingScheme: z.string().nullable(),
    taxBehavior: z.string().nullable(),
    recurringInterval: z.string().nullable(),
    recurringIntervalCount: z.number().nullable(),
    metadata: z.record(z.string()),
    syncedAt: z.string(),
});
export const stripeCustomerSchema = z.object({
    environment: stripeEnvironmentSchema,
    stripeCustomerId: z.string(),
    email: z.string().nullable(),
    name: z.string().nullable(),
    phone: z.string().nullable(),
    deleted: z.boolean(),
    metadata: z.record(z.string()),
    stripeCreatedAt: z.string().nullable(),
    syncedAt: z.string(),
});
export const paymentCustomerListItemSchema = stripeCustomerSchema.extend({
    paymentsCount: z.number().int().nonnegative(),
    lastPaymentAt: z.string().nullable(),
    totalSpend: z.number().int().nonnegative().nullable(),
    totalSpendCurrency: z.string().nullable(),
    paymentMethodBrand: z.string().nullable(),
    paymentMethodLast4: z.string().nullable(),
    countryCode: z.string().trim().length(2).nullable(),
});
export const billingSubjectSchema = z
    .object({
    type: z.string().trim().min(1).max(100),
    id: z.string().trim().min(1).max(255),
})
    .strict();
export const checkoutModeSchema = z.enum(['payment', 'subscription']);
export const checkoutSessionStatusSchema = z.enum([
    'initialized',
    'open',
    'completed',
    'expired',
    'failed',
]);
export const checkoutSessionPaymentStatusSchema = z.enum(['paid', 'unpaid', 'no_payment_required']);
export const checkoutSessionSchema = z.object({
    id: z.string(),
    environment: stripeEnvironmentSchema,
    mode: checkoutModeSchema,
    status: checkoutSessionStatusSchema,
    paymentStatus: checkoutSessionPaymentStatusSchema.nullable(),
    subjectType: z.string().nullable(),
    subjectId: z.string().nullable(),
    customerEmail: z.string().nullable(),
    stripeCheckoutSessionId: z.string().nullable(),
    stripeCustomerId: z.string().nullable(),
    stripePaymentIntentId: z.string().nullable(),
    stripeSubscriptionId: z.string().nullable(),
    url: z.string().nullable(),
    lastError: z.string().nullable(),
    createdAt: z.string(),
    updatedAt: z.string(),
});
export const customerPortalSessionStatusSchema = z.enum(['initialized', 'created', 'failed']);
export const customerPortalSessionSchema = z.object({
    id: z.string(),
    environment: stripeEnvironmentSchema,
    status: customerPortalSessionStatusSchema,
    subjectType: z.string(),
    subjectId: z.string(),
    stripeCustomerId: z.string().nullable(),
    returnUrl: z.string().nullable(),
    configuration: z.string().nullable(),
    url: z.string().nullable(),
    lastError: z.string().nullable(),
    createdAt: z.string(),
    updatedAt: z.string(),
});
export const stripeCustomerMappingSchema = z.object({
    environment: stripeEnvironmentSchema,
    subjectType: z.string(),
    subjectId: z.string(),
    stripeCustomerId: z.string(),
    createdAt: z.string(),
    updatedAt: z.string(),
});
export const paymentHistoryTypeSchema = z.enum([
    'one_time_payment',
    'subscription_invoice',
    'refund',
    'failed_payment',
]);
export const paymentHistoryStatusSchema = z.enum([
    'succeeded',
    'failed',
    'pending',
    'refunded',
    'partially_refunded',
]);
export const paymentHistorySchema = z.object({
    environment: stripeEnvironmentSchema,
    type: paymentHistoryTypeSchema,
    status: paymentHistoryStatusSchema,
    subjectType: z.string().nullable(),
    subjectId: z.string().nullable(),
    stripeCustomerId: z.string().nullable(),
    customerEmailSnapshot: z.string().nullable(),
    stripeCheckoutSessionId: z.string().nullable(),
    stripePaymentIntentId: z.string().nullable(),
    stripeInvoiceId: z.string().nullable(),
    stripeChargeId: z.string().nullable(),
    stripeRefundId: z.string().nullable(),
    stripeSubscriptionId: z.string().nullable(),
    stripeProductId: z.string().nullable(),
    stripePriceId: z.string().nullable(),
    amount: z.number().nullable(),
    amountRefunded: z.number().nullable(),
    currency: z.string().nullable(),
    description: z.string().nullable(),
    paidAt: z.string().nullable(),
    failedAt: z.string().nullable(),
    refundedAt: z.string().nullable(),
    stripeCreatedAt: z.string().nullable(),
    createdAt: z.string(),
    updatedAt: z.string(),
});
export const stripeSubscriptionStatusSchema = z.enum([
    'incomplete',
    'incomplete_expired',
    'trialing',
    'active',
    'past_due',
    'canceled',
    'unpaid',
    'paused',
]);
export const stripeSubscriptionItemSchema = z.object({
    environment: stripeEnvironmentSchema,
    stripeSubscriptionItemId: z.string(),
    stripeSubscriptionId: z.string(),
    stripeProductId: z.string().nullable(),
    stripePriceId: z.string().nullable(),
    quantity: z.number().nullable(),
    metadata: z.record(z.string()),
    createdAt: z.string(),
    updatedAt: z.string(),
});
export const stripeSubscriptionSchema = z.object({
    environment: stripeEnvironmentSchema,
    stripeSubscriptionId: z.string(),
    stripeCustomerId: z.string(),
    subjectType: z.string().nullable(),
    subjectId: z.string().nullable(),
    status: stripeSubscriptionStatusSchema,
    currentPeriodStart: z.string().nullable(),
    currentPeriodEnd: z.string().nullable(),
    cancelAtPeriodEnd: z.boolean(),
    cancelAt: z.string().nullable(),
    canceledAt: z.string().nullable(),
    trialStart: z.string().nullable(),
    trialEnd: z.string().nullable(),
    latestInvoiceId: z.string().nullable(),
    metadata: z.record(z.string()),
    syncedAt: z.string(),
    createdAt: z.string(),
    updatedAt: z.string(),
    items: z.array(stripeSubscriptionItemSchema).optional(),
});
export const stripeWebhookProcessingStatusSchema = z.enum([
    'pending',
    'processed',
    'failed',
    'ignored',
]);
export const stripeWebhookEventSchema = z.object({
    environment: stripeEnvironmentSchema,
    stripeEventId: z.string(),
    eventType: z.string(),
    livemode: z.boolean(),
    stripeAccountId: z.string().nullable(),
    objectType: z.string().nullable(),
    objectId: z.string().nullable(),
    processingStatus: stripeWebhookProcessingStatusSchema,
    attemptCount: z.number(),
    lastError: z.string().nullable(),
    receivedAt: z.string(),
    processedAt: z.string().nullable(),
    createdAt: z.string(),
    updatedAt: z.string(),
});
