// @insforge/shared-schemas/dist/posthog.schema.js
import { z } from 'zod';
export const posthogConnectionStatusSchema = z.enum(['active', 'degraded', 'revoked']);
export const posthogConnectionSchema = z.object({
    posthogProjectId: z.string(),
    organizationName: z.string().nullable(),
    projectName: z.string(),
    region: z.enum(['US', 'EU']),
    host: z.string().url(),
    apiKey: z.string(),
    status: posthogConnectionStatusSchema,
    createdAt: z.string(),
});
export const posthogDashboardSchema = z.object({
    id: z.number(),
    name: z.string(),
    description: z.string().optional(),
    pinned: z.boolean().optional(),
    lastModifiedAt: z.string().optional(),
    url: z.string().url(),
});
export const posthogDashboardsResponseSchema = z.object({
    dashboards: z.array(posthogDashboardSchema),
    count: z.number(),
});
export const posthogSummarySchema = z.object({
    todayEvents: z.number(),
    dau24h: z.number(),
    totalEvents7d: z.number(),
    topEvents: z.array(z.object({ event: z.string(), count: z.number() })),
});
export const posthogEventRecordSchema = z.object({
    id: z.string(),
    event: z.string(),
    distinctId: z.string(),
    timestamp: z.string(),
});
export const posthogEventsResponseSchema = z.object({
    events: z.array(posthogEventRecordSchema),
    next: z.string().nullable(),
});
// v2.5 dashboard ----------------------------------------------------------
export const posthogTimeframeSchema = z.enum(['24h', '7d', '30d', '3m']);
export const posthogBreakdownSchema = z.enum(['Page', 'Country', 'DeviceType']);
export const posthogMetricSchema = z.enum(['visitors', 'views', 'bounce_rate']);
export const posthogWebOverviewItemSchema = z.object({
    key: z.string(),
    value: z.number().nullable(),
    previous: z.number().nullable(),
    changeFromPreviousPct: z.number().nullable(),
    isIncreaseBad: z.boolean().nullable().optional(),
});
export const posthogWebOverviewResponseSchema = z.object({
    items: z.array(posthogWebOverviewItemSchema),
});
export const posthogWebStatsRowSchema = z.object({
    breakdownValue: z.string().nullable(),
    visitors: z.number(),
    views: z.number(),
    uiFillFraction: z.number(),
});
export const posthogWebStatsResponseSchema = z.object({
    rows: z.array(posthogWebStatsRowSchema),
});
export const posthogTrendPointSchema = z.object({
    date: z.string(),
    count: z.number(),
});
export const posthogTrendsResponseSchema = z.object({
    series: z.array(posthogTrendPointSchema),
});
export const posthogRetentionRowSchema = z.object({
    date: z.string(),
    label: z.string(),
    values: z.array(z.object({ count: z.number().nullable() })),
});
export const posthogRetentionResponseSchema = z.object({
    rows: z.array(posthogRetentionRowSchema),
});
export const posthogRecordingItemSchema = z.object({
    id: z.string(),
    distinctId: z.string(),
    durationSeconds: z.number(),
    startTime: z.string(),
    endTime: z.string(),
    startUrl: z.string().nullable(),
    clickCount: z.number(),
    consoleErrorCount: z.number(),
});
export const posthogRecordingsResponseSchema = z.object({
    items: z.array(posthogRecordingItemSchema),
});
// `z.string().url()` only checks syntactic validity. We render this URL inside
// an iframe with `allow-same-origin`, so restrict the origin to PostHog hosts.
// Regex (vs `new URL`) because shared-schemas builds without DOM lib globals.
const posthogUrlPattern = /^https:\/\/([a-z0-9-]+\.)*posthog\.com(?::\d+)?(\/|$)/i;
export const posthogShareTokenResponseSchema = z.object({
    embedUrl: z
        .string()
        .url()
        .refine((u) => posthogUrlPattern.test(u), {
        message: 'embedUrl must be an https://*.posthog.com URL',
    }),
});
